apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfSubscriber.cs

```diff
@@ -0,0 +1,723 @@
+        /// <param name="shutdown">Function to call on this <see cref="HiPerfSubscriberImpl"/> shutdown.</param>
+        public HiPerfSubscriberImpl(SubscriptionName subscriptionName, IEnumerable<SubscriberClient> clients, Settings settings, Func<Task> shutdown)
+            : this(subscriptionName, clients, settings, shutdown, TaskHelper.Default) { }
+
+        internal HiPerfSubscriberImpl(SubscriptionName subscriptionName, IEnumerable<SubscriberClient> clients, Settings settings, Func<Task> shutdown, TaskHelper taskHelper)
+        {
+            _subscriptionName = GaxPreconditions.CheckNotNull(subscriptionName, nameof(subscriptionName));
+            GaxPreconditions.CheckNotNull(clients, nameof(clients));
+            _clients = clients.ToArray();
+            GaxPreconditions.CheckArgument(_clients.Length > 0, nameof(clients), "Must contain at least one client");
+            GaxPreconditions.CheckArgument(_clients.All(x => x != null), nameof(clients), "All elements must be non-null");
+            GaxPreconditions.CheckNotNull(settings, nameof(settings));
+            settings.Validate();
+            _modifyDeadlineSeconds = (int)((settings.StreamAckDeadline ?? DefaultStreamAckDeadline).TotalSeconds);
+            _autoExtendInterval = TimeSpan.FromSeconds(_modifyDeadlineSeconds) - settings.AckExtensionWindow ?? DefaultAckExtensionWindow;
+            _maxAckExtendCount = 10_000;
+            _shutdown = shutdown;
+            _scheduler = settings.Scheduler;
+            _taskHelper = GaxPreconditions.CheckNotNull(taskHelper, nameof(taskHelper));
+            _flowControlSettings = settings.FlowControlSettings ?? DefaultFlowControlSettings;
+        }
+
+        private readonly object _lock = new object();
+        private readonly SubscriptionName _subscriptionName;
+        private readonly SubscriberClient[] _clients;
+        private readonly Func<Task> _shutdown;
+        private readonly TimeSpan _autoExtendInterval; // Interval between message lease auto-extends
+        private readonly int _modifyDeadlineSeconds; // Value to use as new deadline when lease auto-extends
+        private readonly int _maxAckExtendCount; // Maximum count of acks/extends to push to server in a single messages
+        private readonly IScheduler _scheduler;
+        private readonly TaskHelper _taskHelper;
+        private readonly FlowControlSettings _flowControlSettings;
+
+        private TaskCompletionSource<int> _mainTcs;
+        private CancellationTokenSource _globalSoftStopCts;
+        private CancellationTokenSource _globalHardStopCts;
+
+        /// <inheritdoc />
+        public override Task StartAsync(Func<PubsubMessage, CancellationToken, Task<Reply>> handlerAsync)
+        {
+            lock (_lock)
+            {
+                GaxPreconditions.CheckState(_mainTcs == null, "Can only start an instance once.");
```

Comment:
Gosh, you can't even restart it after shutting it down? That's a bit surprising, particularly if it's creating a load of clients. Maybe it fits in with the common use case though.